File: src/lib/scripts/generators/generate_component.py
import argparse
import sys
from termcolor import colored
from lib.scripts.utils.helpers import *

# Constants
COMPONENT_EXISTS = 'That component already exists.'

def parse_args(args):
    # Define + use parser
    parser = argparse.ArgumentParser(description='Create a new component.')
    parser.add_argument('subdir', metavar='SUBDIRECTORY_NAME', type=str, nargs=1,
                        help='the subdirectory of the component')
    parser.add_argument('component', metavar='COMPONENT_NAME', type=str, nargs=1,
                        help='the name of the component in PascalCase')
    return parser.parse_args(args)


def get_paths(args):
    subdir_path = f'src/components/{args.subdir[0]}'
    component_path = subdir_path + '/' + args.component[0]
    return (subdir_path, component_path)


def validate_args(args):
    # Check for existing paths
    subdir_path, component_path = get_paths(args)
    if not os.path.isdir(subdir_path):
        raise Exception(SUBDIRECTORY_DOESNT_EXIST)
    if os.path.isdir(component_path):
        raise Exception(COMPONENT_EXISTS)


def create_tsx(path, component):
    kebab_component = to_kebab_case(component)
    file_path = f'{path}/{component}.tsx'
    file_contents = [
        f'import styles from "./{component}.module.scss";\n',
        '\n',
        f'const {component}: React.FC = () => {{\n',
        DIVIDER_STRING,
        f'\t// MARKUP\n',
        DIVIDER_STRING,
        '\n',
        f'\treturn <div className={{styles["{kebab_component}"]}}></div>;\n',
        f'}};\n',
        '\n',
        f'export default {component};\n'
    ]

    # Handle single-word components
    if '-' not in kebab_component:
        idx = file_contents.index(f'\treturn <div className={{styles["{kebab_component}"]}}></div>;\n')
        file_contents[idx] = f'\treturn <div className={{styles.{kebab_component}}}></div>;\n'

    write_to_file(file_path, file_contents)


def create_scss(path, component):
    file_path = f'{path}/{component}.module.scss'
    file_contents = [
        '@use "@/styles/common" as *;\n',
        '\n',
        f'.{to_kebab_case(component)} {{\n',
        '\n',
        '}\n'
    ]

    write_to_file(file_path, file_contents)


def create_md(path, component):
    file_path = f'{path}/{component}.md'
    write_to_file(file_path, [])


def create_component(args):
    _, component_path = get_paths(args)

    # Create component folder
    os.makedirs(component_path)

    # Create files
    create_tsx(component_path, args.component[0])
    create_scss(component_path, args.component[0])
    create_md(component_path, args.component[0])

    print(colored(f'{args.component[0]} was created successfully!', GREEN))


def new():
    try:
        validate_root_directory()
    except Exception as e:
        handle_exception(e)
    
    args = parse_args(sys.argv[1:])

    try:
        validate_args(args)
    except Exception as e:
        handle_exception(e)

    create_component(args)


if __name__ == '__main__':
    new()

File: src/lib/scripts/generators/generateComponent.ts
import fs from 'fs';
import path from 'path';
import {
    DIVIDER_STRING,
    handleException,
    logSuccess,
    SUBDIRECTORY_DOESNT_EXIST,
    toKebabCase,
    validateRootDirectory,
    writeToFile,
} from '@/lib/scripts/utils/helpers';

const COMPONENT_EXISTS = 'That component already exists.';
const USAGE = 'Usage: npm run gen:component <subdirectory> <ComponentName>';

interface ComponentArgs {
    subdir: string;
    component: string;
}

const parseArgs = (argv: string[]): ComponentArgs => {
    const [subdir, component] = argv;
    if (!subdir || !component) {
        throw new Error(USAGE);
    }
    return { subdir, component };
};

const getPaths = (args: ComponentArgs) => {
    const subdirPath = path.join('src', 'components', args.subdir);
    const componentPath = path.join(subdirPath, args.component);
    return { subdirPath, componentPath };
};

const validateArgs = (args: ComponentArgs): void => {
    const { subdirPath, componentPath } = getPaths(args);
    if (!fs.existsSync(subdirPath) || !fs.statSync(subdirPath).isDirectory()) {
        throw new Error(SUBDIRECTORY_DOESNT_EXIST);
    }
    if (fs.existsSync(componentPath)) {
        throw new Error(COMPONENT_EXISTS);
    }
};

const createTsx = (componentPath: string, component: string): void => {
    const kebabComponent = toKebabCase(component);
    const filePath = path.join(componentPath, `${component}.tsx`);
    const classNameAccess = kebabComponent.includes('-')
        ? `styles['${kebabComponent}']`
        : `styles.${kebabComponent}`;

    writeToFile(filePath, [
        `import styles from './${component}.module.scss';\n`,
        '\n',
        `const ${component}: React.FC = () => {\n`,
        DIVIDER_STRING,
        '\t// MARKUP\n',
        DIVIDER_STRING,
        '\n',
        `\treturn <div className={${classNameAccess}}></div>;\n`,
        '};\n',
        '\n',
        `export default ${component};\n`,
    ]);
};

const createScss = (componentPath: string, component: string): void => {
    const filePath = path.join(componentPath, `${component}.module.scss`);

    writeToFile(filePath, [
        "@use '@/styles/constants' as *;\n",
        "@use '@/styles/mixins' as *;\n",
        '\n',
        `.${toKebabCase(component)} {\n`,
        '}\n',
    ]);
};

const createMd = (componentPath: string, component: string): void => {
    const filePath = path.join(componentPath, `${component}.md`);
    writeToFile(filePath, [`# ${component}\n`]);
};

const createComponent = (args: ComponentArgs): void => {
    const { componentPath } = getPaths(args);

    fs.mkdirSync(componentPath, { recursive: true });

    createTsx(componentPath, args.component);
    createScss(componentPath, args.component);
    createMd(componentPath, args.component);

    logSuccess(`${args.component} was created successfully!`);
};

const main = (): void => {
    try {
        validateRootDirectory();
        const args = parseArgs(process.argv.slice(2));
        validateArgs(args);
        createComponent(args);
    } catch (error) {
        handleException(error);
    }
};

main();
